ibis/expr/datatypes.py

```diff
@@ -1103,17 +1163,20 @@ def higher_precedence(left, right):
     )
 
 
-def highest_precedence(dtypes):
+def highest_precedence(dtypes: Iterator[DataType]) -> DataType:
+    """Compute the highest precedence of `dtypes`."""
     return functools.reduce(higher_precedence, dtypes)
 
 
 @infer.register(object)
-def infer_dtype_default(value):
+def infer_dtype_default(value: GenericAny) -> DataType:
+    """Default implementation of :func:`~ibis.expr.datatypes.infer`."""
     raise com.InputTypeError(value)
 
 
 @infer.register(collections.OrderedDict)
-def infer_struct(value):
+def infer_struct(value: collections.OrderedDict[str, GenericAny]) -> Struct:
```

Comment:
```suggestion
def infer_struct(value: collections.OrderedDict[str, Union[str, DataType]]) -> Struct:
```

Perhaps introduce a typehint alias, something like `DTypeLike = Union[str, DataType]`